File: push.py
#!/usr/bin/env python
"""FreeSWITCH Push.

Send GCM/Apple Push/Windows Phone Push notifications to mobile softphones.
"""
import sys
import requests
import ESL
import json
con = ESL.ESLconnection("127.0.0.1", "8021", "ClueCon")


keys = json.load(open("/etc/pushkeys.conf"))


def parse_contact(unparse):
    """parse_reg(contact): parses FreeSWITCH registration strings.

    Parameters:
        * contact: a single contact string. The sofia_contact function returns a comma separated
                   list of all of the contact strings for a given extension
    Returns:
        a dict of all of the variables in the contact string, or False if the string is an error
"""
    components = unparse.split(";")
    c = components[0].split("/", 2)
    if c[0] == "error":
        return False
    args = {}
    for component in components[1:]:
        if "=" in component:
            key, value = component.strip().split("=", 2)
            args[key] = value
        else:
            args[component.strip()] = None
    out = {
        "args": args,
        "profile": None,
        "uri": None
    }
    if len(c) > 1:
        out["profile"] = c[1]
    if len(c) > 2:
        out["uri"] = c[2]
    return out


def zoiper(url):
    """Send a Zoiper push notification to a WP8 device."""
    headers = {"Content-Type": "text/xml", "X-NotificationClass": 4}
    postdata = """<?xml version="1.0" encoding="utf-8"?><root><Value1>Zoiper</Value1><Value2>Incoming</Value2><Value3>Call</Value3></root>"""
    requests.post(url, data=postdata, headers=headers)


def gcm(key, token, data=None):
    """Send a GCM push notification."""
    url = "https://gcm-http.googleapis.com/gcm/send"
    headers = {
        "Authorization": "key=%s" % keys['gcm'][key],
        "Content-Type": "application/json"
    }
    postdata = {"to": token}
    if data is not None:
        postdata['data'] = data
    requests.post(url, data=json.dumps(postdata), headers=headers)


def linphone(args):
    """Trigger a push notification for Linphone."""
    if args['pn-type'] == "google":
        if args['app-id'] in keys['gcm']:
            gcm(args['app-id'], args['pn-tok'])

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: %s [action] [extension@realm]" % sys.argv[0])
        print("\nThe following actions are supported:")
        print(" * push - checks all of the contact strings stored for the extension and sends them push notifications if applicable")
        print(" * status - displays all of the contact strings, mostly for debugging")
    else:
        action = sys.argv[1]
        contacts = con.api("sofia_contact %s" % sys.argv[2]).getBody().split(",")
        if action == "push":
            delay = 0
            for contact in contacts:
                c = parse_contact(contact)
                if c:
                    if "X-PUSH-URI" in c['args']:
                        zoiper(c['args']['X-PUSH-URI'])
                        if delay < 2000:
                            delay = 2000
                    elif "pn-type" in c['args']:
                        linphone(c['args'])
                        if delay < 2000:
                            delay = 2000
            print(str(delay))
        elif action == "status":
            for contact in contacts:
                c = parse_contact(contact)
                if c:
                    print("\n%s on profile %s" % (c['uri'], c['profile']))
                    for i in c['args']:
                        print(" * %s:\t%s" % (i, c['args'][i]))
                else:
                    print("No contact found")

File: push.lua
function split (inputstr, sep)
	if (sep == nil) then
		sep = "%s"
	end
	local t={}
	for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
		table.insert(t, str)
	end
	return t
end

function len (array)
	count = 0
	for Index, Value in pairs(array) do
		count = count + 1
	end
	return count
end


function parse_contact (unparsed)
	components = split(unparsed, ";")
	c = split(components[1], "/")
	if (c[1] == "error") then
		return nil
	end
	args = {}
	for i, component in ipairs(components) do
		if (i == 0) then
		elseif (string.find(component,"=")) then
			parts = split(component, "=")
			args[parts[1]] = parts[2]
		end
	end
	out = {
		args = args,
		profile = nil,
		uri = nil
	}
--[[	if (len(c) > 1) then
		if (len(c) > 2) then
			out[uri] = c[2]
		end
		out[profile] = c[1]
	end]]-- disabled due to errors and not needed
	return out
end

sentCount=0
api = freeswitch.API();
local fb = require('firebase'):new({PROJECT_FN = "/var/firebase/firebase-session.json", PROJECT_ID="accelerate-networks"})                   
fb.auth:auth_service_account('/etc/firebase/google-account.json')
contacts = split(api:executeString("sofia_contact " .. session:getVariable("destination_number") .. "@" .. session:getVariable("domain_name")), ",")
if contacts then
	for i, contact_info in pairs(contacts) do
		delay = 0
		contact = parse_contact(contact_info)
		if (contact) then
			if (contact["args"] and contact["args"]["pn-type"] and contact["args"]["pn-type"] == "firebase" and contact["args"]["pn-tok"]) then
				fb.messaging:send({
					message = {
						token = contact["args"]["pn-tok"],                                                           
						android = {
							priority = "HIGH"                                                                    
						}
					}
				})
				sentCount = sentCount + 1
				if delay < 2000 then
					delay = 2000
				end
			end
		end
	end
end
session:execute("sleep", delay)
session:consoleLog("info", "Sent push notifaction to " .. sentCount .. " device(s)") 
